File: green-campus-web-mobile-main/green-campus-web-mobile-main/mobile/lib/data/services/profile_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:http/io_client.dart';
import 'package:mobile_app/core/config/app_config.dart';
import 'package:mobile_app/data/services/token_storage.dart';

class ProfileService {
  late final http.Client _client;

  ProfileService() {
    final ioClient = HttpClient();
    ioClient.badCertificateCallback =
        (X509Certificate cert, String host, int port) {
          return host == '10.0.2.2' || host == 'localhost';
        };
    _client = IOClient(ioClient);
  }

  /// Actualiza nombre, apellido y contraseña del usuario autenticado.
  /// La contraseña es opcional: si es null o vacía, no se envía al servidor.
  Future<http.Response> updateProfile({
    required String firstName,
    required String lastName,
    String? password,
  }) async {
    final token = await TokenStorage.getToken();

    final body = <String, dynamic>{
      'firstName': firstName,
      'lastName': lastName,
    };
    if (password != null && password.isNotEmpty) {
      body['password'] = password;
    }

    return await _client
        .patch(
          Uri.parse(AppConfig.profileEndpoint),
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'Authorization': 'Bearer $token',
          },
          body: jsonEncode(body),
        )
        .timeout(const Duration(seconds: 10));
  }
}

File: green-campus-web-mobile-main/green-campus-web-mobile-main/web/src/app/data/services/profile.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { APP_CONFIG } from '../../core/config/app-config';
import { TokenStorageService } from './token-storage.service';

export interface UpdateProfileRequest {
  firstName?: string;
  lastName?: string;
  password?: string;
}

@Injectable({ providedIn: 'root' })
export class ProfileService {
  private http = inject(HttpClient);
  private tokenStorage = inject(TokenStorageService);
  private base = APP_CONFIG.apiBaseUrl;

  private get headers(): HttpHeaders {
    return new HttpHeaders({ Authorization: `Bearer ${this.tokenStorage.getJwt()}` });
  }

  updateProfile(data: UpdateProfileRequest): Observable<void> {
    return this.http
      .patch<void>(`${this.base}/user/profile`, data, { headers: this.headers })
      .pipe(
        tap(() => {
          if (data.firstName !== undefined && data.lastName !== undefined) {
            this.tokenStorage.updateProfile(data.firstName, data.lastName);
          }
        }),
      );
  }
}
